Bugfix FXIOS-11421 [UI] Fix ButtonToast alignment issue

VirtualPixel opened the pull request:
## :scroll: Tickets
[Jira ticket](https://mozilla-hub.atlassian.net/browse/FXIOS-11421)
[Github issue](https://github.com/mozilla-mobile/firefox-ios/issues/24851)

## :bulb: Description
I fixed an alignment issue in the ButtonToast component where the checkmark icon was displaying above the title instead of to the left of it. The solution:

- Moved button from contentStackView to labelStackView to ensure proper vertical alignment, this ensures the button remains below the title/description text
- Updated adjustLayoutForA11ySizeCategory to properly handle stack orientation, this also keeps the checkmark centered vertically
- Ensured button appears below text content rather than beside it
- Maintained proper accessibility support with dynamic type sizes
- Ensured padding is present

This aligns ButtonToast more closely with how DownloadToast is set up.

<table>
  <tr>
    <th>Orientation</th>
    <th>Before</th>
    <th>After</th>
  </tr>
  <tr>
    <td><strong>Portrait</strong></td>
    <td><img src="https://github.com/user-attachments/assets/c8afceb3-59c7-41ab-91c9-db9ea7ea1b3f" width="150" alt="Before - Portrait"></td>
    <td><img src="https://github.com/user-attachments/assets/926374b1-2228-43d2-8d97-4b86093b558a" width="150" alt="After - Portrait"></td>
  </tr>

  <tr>
    <td><strong>Landscape</strong></td>
    <td><img src="https://github.com/user-attachments/assets/e1d02510-0530-45da-ac84-3c95f016bb9e" width="450" alt="Before - Landscape"></td>
    <td><img src="https://github.com/user-attachments/assets/b39bec06-e6c1-40a3-bed4-671adaac490a" width="450" alt="After - Landscape"></td>
  </tr>
</table>

## :pencil: Checklist
You have to check all boxes before merging
- [X] Filled in the above information (tickets numbers and description of your work)
- [X] Updated the PR name to follow our [PR naming guidelines](https://github.com/mozilla-mobile/firefox-ios/wiki/Pull-Request-Naming-Guide)
- [X] Wrote unit tests and/or ensured the tests suite is passing
- [X] When working on UI, I checked and implemented accessibility (minimum Dynamic Text and VoiceOver)
- [X] If needed, I updated documentation / comments for complex code and public methods
- [ ] If needed, added a backport comment (example `@Mergifyio backport release/v120`)

VirtualPixel commented:
Hey, no problem at all 😊 I'll make the updates shortly and I'll reply again once it's done. Thanks for updating the ticket

FilippoZazzeroni commented:
Thanks very much @VirtualPixel

VirtualPixel commented:
> Thanks very much @VirtualPixel

@FilippoZazzeroni  The updated code has been committed to the pull request, I've also attached screenshots of before/after for the toast 😊

FilippoZazzeroni commented on firefox-ios/Client/Frontend/Browser/ButtonToast.swift:
```diff
@@ -108,15 +111,39 @@ class ButtonToast: Toast {
         ])
+
+        if let buttonText = viewModel.buttonText {
+            roundedButton.setTitle(buttonText, for: [])
```
this whole if else it should be deleted IMO since, setupPaddedButton is already adding the button to a view and also adding the constraints to it. 
i'd refactor the `setupPaddedButton` to remove all the constraints on the intrinsic content size since the button already does that. Also we should remove the `addGestureRecognizer` in the `setupPaddedButton` since buttons already can handle tap, we can just add an action via `button.addAction`.

FilippoZazzeroni commented on firefox-ios/Client/Frontend/Browser/ButtonToast.swift:
```diff
@@ -108,15 +111,39 @@ class ButtonToast: Toast {
         NSLayoutConstraint.activate([
             contentStackView.leadingAnchor.constraint(equalTo: toastView.leadingAnchor, constant: UX.spacing),
             contentStackView.trailingAnchor.constraint(equalTo: toastView.trailingAnchor, constant: -UX.spacing),
-            contentStackView.bottomAnchor.constraint(equalTo: toastView.bottomAnchor, constant: -UX.spacing),
```
this should be added back

FilippoZazzeroni commented on firefox-ios/Client/Frontend/Browser/ButtonToast.swift:
```diff
@@ -149,13 +176,13 @@ class ButtonToast: Toast {
-            contentStackView.distribution = .fillProportionally
+            // Description label changes with progress and if this isn't clipped the height of the
+            // toast changes continually while loading
+            descriptionLabel.numberOfLines = 1
```
CC @thatswinnie. I don't think we should limit the number of lines to 1, otherwise we can't read the text.

FilippoZazzeroni reviewed (changes requested):
I added some comments to the PR, let me know if something is not clear or doesn't make sense. Also i see it is missing the safe area constraint, we should constraint the toast to leading and trailing safe area otherwise it goes underneath the dynamic island or the notch

VirtualPixel commented:
> I added some comments to the PR, let me know if something is not clear or doesn't make sense. Also i see it is missing the safe area constraint, we should constraint the toast to leading and trailing safe area otherwise it goes underneath the dynamic island or the notch

I agree, I added the constrains from:
`toastView.leadingAnchor.constraint(equalTo: leadingAnchor, constant: Toast.UX.shadowVerticalSpacing),
            toastView.trailingAnchor.constraint(equalTo: trailingAnchor, constant: -Toast.UX.shadowVerticalSpacing),`

to this: 
`NSLayoutConstraint.activate([
            toastView.leadingAnchor.constraint(equalTo: safeAreaLayoutGuide.leadingAnchor,
                                               constant: Toast.UX.shadowVerticalSpacing),
            toastView.trailingAnchor.constraint(equalTo: safeAreaLayoutGuide.trailingAnchor,
                                                constant: -Toast.UX.shadowVerticalSpacing),
            toastView.heightAnchor.constraint(equalTo: heightAnchor,
                                              constant: -Toast.UX.shadowHorizontalSpacing),

            heightAnchor.constraint(greaterThanOrEqualToConstant: Toast.UX.toastHeightWithShadow)
        ])`
        
Though this does look a little strange since the "DownloadToast" that shows before the ButtonToast doesn't follow the constraints and goes under the dynamic island as well. Should this be another ticket, or do you want me to include the fix for that in here with it?

VirtualPixel commented:
@FilippoZazzeroni New commit is in, I also synced the branches so it's up to date. Let me know if you have any questions 😊

FilippoZazzeroni commented on firefox-ios/Client/Frontend/Browser/ButtonToast.swift:
```diff
@@ -108,32 +113,43 @@ class ButtonToast: Toast {
 
-        stackView.addArrangedSubview(roundedButton)
         roundedButton.setTitle(buttonText, for: [])
+        roundedButton.translatesAutoresizingMaskIntoConstraints = false
```
this is not needed since the `.build()` method sets it already

VirtualPixel replied:
Fair! I removed it now 😊

FilippoZazzeroni commented on firefox-ios/Client/Frontend/Browser/ButtonToast.swift:
```diff
@@ -108,32 +113,43 @@ class ButtonToast: Toast {
-            roundedButton.widthAnchor.constraint(
-                equalToConstant: roundedButton.titleLabel!.intrinsicContentSize.width + 2 * UX.buttonPadding)
+            // Minimum size constraints
+            roundedButton.heightAnchor.constraint(greaterThanOrEqualTo: roundedButton.titleLabel!.heightAnchor,
```
this should be removed, since the button should size directly to its intrinsic size

VirtualPixel replied:
This is true, it also added padding to the inside of the button so I will remove this and add padding through content insets instead

FilippoZazzeroni commented on firefox-ios/Client/Frontend/Browser/ButtonToast.swift:
```diff
@@ -108,32 +113,43 @@ class ButtonToast: Toast {
+            // Position constraints 
+            roundedButton.leadingAnchor.constraint(equalTo: labelStackView.leadingAnchor),
+            roundedButton.topAnchor.constraint(equalTo: stackView.bottomAnchor, constant: UX.spacing),
+            roundedButton.bottomAnchor.constraint(equalTo: toastView.bottomAnchor, constant: -UX.spacing)
```
We should add also the trailing constraint so the button doesn't go after the toast. 
It should be something like:
```swift
roundedButton.trailingAnchor.constraint(greaterThanOrEqualTo: toastView.leadingAnchor, constant: -UX.spacing)
```

VirtualPixel replied:
Added 😊

FilippoZazzeroni commented on firefox-ios/Client/Frontend/Browser/ButtonToast.swift:
```diff
@@ -149,13 +165,13 @@ class ButtonToast: Toast {
-            contentStackView.axis = .vertical
-            contentStackView.alignment = .leading
-            contentStackView.distribution = .fillProportionally
+            // Description label changes with progress and if this isn't clipped the height of the
```
there is no progress here so this it should not be needed, also i'm not sure but if the lines of description label are 0 why do we need line break mode, we should not have any wrapping happening. Tagging @thatswinnie for extra review on accessibility 😃

VirtualPixel replied:
I've reworked the code based on your feedback:

- Put back the original contentStackView config
- Checking horizontalSizeClass.regular instead of landscape per @thatswinnie's suggestion
- Going with 2 lines for the description - seems like a good middle ground
- Using truncation to keep the height stable

What do you think? Is 2 lines the right balance?

FilippoZazzeroni commented on firefox-ios/Client/Frontend/Browser/ButtonToast.swift:
```diff
@@ -108,32 +113,43 @@ class ButtonToast: Toast {
+            // Minimum size constraints
+            roundedButton.heightAnchor.constraint(greaterThanOrEqualTo: roundedButton.titleLabel!.heightAnchor,
+                                                  constant: UX.buttonPadding),
+            roundedButton.widthAnchor.constraint(greaterThanOrEqualTo: roundedButton.titleLabel!.widthAnchor,
```
same

VirtualPixel replied:
Removed like the other 😊

FilippoZazzeroni reviewed (changes requested):
Hey @VirtualPixel thanks for the updates  😃  i added some comments to the implementation.

VirtualPixel commented on firefox-ios/Client/Frontend/Browser/ButtonToast.swift:
```diff
@@ -108,32 +113,43 @@ class ButtonToast: Toast {
             contentStackView.trailingAnchor.constraint(equalTo: toastView.trailingAnchor, constant: -UX.spacing),
-            contentStackView.bottomAnchor.constraint(equalTo: toastView.bottomAnchor, constant: -UX.spacing),
-            contentStackView.topAnchor.constraint(equalTo: toastView.topAnchor, constant: UX.spacing)
+            contentStackView.bottomAnchor.constraint(lessThanOrEqualTo: toastView.bottomAnchor, constant: -UX.spacing),
```
I tried adding the constraint back with equalTo as recommended, but it was still cutting off the bottom of the button text. Using lessThanOrEqualTo fixed this issue while maintaining the proper spacing control. It prevents the layout conflict between the two bottom constraints

VirtualPixel commented:
> Hey @VirtualPixel thanks for the updates  😃  i added some comments to the implementation. 

@FilippoZazzeroni Its ready for review when you're able, no rush though 😊

FilippoZazzeroni commented on firefox-ios/Client/Frontend/Browser/ButtonToast.swift:
```diff
@@ -108,32 +113,44 @@ class ButtonToast: Toast {
 
-        stackView.addArrangedSubview(roundedButton)
         roundedButton.setTitle(buttonText, for: [])
+        roundedButton.configuration = UIButton.Configuration.plain()
```
Could we set the title trough the configuration too ? so we use the new button API

VirtualPixel replied:
Good catch! I'll update that and comment once I've made the change 😊

FilippoZazzeroni commented on firefox-ios/Client/Frontend/Browser/ButtonToast.swift:
```diff
@@ -147,16 +164,8 @@ class ButtonToast: Toast {
-            contentStackView.alignment = .center
-            contentStackView.distribution = .fill
-        }
+        descriptionLabel.numberOfLines = 3
```
I think we should aim for 0 here cc: @thatswinnie

VirtualPixel replied:
0 would work too, the idea behind 3 was so that the button wouldn't have the ability to expand across the whole screen.

If this isn't a problem, I'll gladly set to 0 😊

FilippoZazzeroni reviewed (commented):
@VirtualPixel great work 🎉  we are almost there, just some minor adjustments

VirtualPixel commented:
@FilippoZazzeroni @thatswinnie I've addressed the comments again, as well as updated the branch 😊

If there's anything else that needs done, feel free to let me know!

VirtualPixel commented on firefox-ios/Client/Frontend/Browser/ButtonToast.swift:
```diff
@@ -142,21 +161,18 @@ class ButtonToast: Toast {
-            contentStackView.distribution = .fill
-        }
+        descriptionLabel.numberOfLines = 0
+        descriptionLabel.lineBreakMode = .byTruncatingTail
```
@FilippoZazzeroni Looking it back over, this is true. Since it's initializing as 0, with the new setup we don't have any real reason to change it. I will delete this and then move the lineBreakMode into the initialization as well

VirtualPixel commented on firefox-ios/Client/Frontend/Browser/ButtonToast.swift:
```diff
@@ -142,21 +161,18 @@ class ButtonToast: Toast {
         imageView.tintColor = theme.colors.textInverted
-        roundedButton.setTitleColor(theme.colors.textInverted, for: [])
+
+        if var buttonConfig = roundedButton.configuration {
```
True, sorry I usually explicitly unwrap these. This is updated

FilippoZazzeroni replied:
no problem

VirtualPixel commented:
@FilippoZazzeroni Updated 😊

FilippoZazzeroni reviewed (approved):
Looks good to me @VirtualPixel there is only a nit about the description label i don't think we need the line break mode any more since with 0 lines there wouldn't be a break behavior occuring.

VirtualPixel commented:
> Hi @VirtualPixel it seems some UI related test are failing, could you please check locally by running the UI test ?

Sure! I'll let you know how it goes, I'm starting the UI testing now

VirtualPixel commented:
Fixed the local testing and should be good now @FilippoZazzeroni @thatswinnie 

Sorry for the delay in the fix!
